Adding new network plugin

bhariharan opened the pull request:
`forcetk` changes coming up next. Do not merge until then.

wmathurin commented on gen/plugins/com.salesforce/com.salesforce.plugin.network.js:
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2012-present, salesforce.com, inc.
```
2016 maybe? ;-)

bhariharan replied:
That's because of the `split` script. The big `cordova_plugins.js` has `2012-present`, and the script splits them up.

wmathurin replied:
Ah that's true

wmathurin commented:
react.force.js includes sendRequest along with all the methods that currently live in forcetk.mobilesdk.js - should we do the same for cordova.force.js?

Or conversely when you refactor force.js should it be such that it can be used from any context (react / cordova or web) maybe through the use of an init method that would take in the function to use to send requests?

bhariharan commented:
I was going for minimal changes in `forcetk` initially. I'm adding a new constructor in `forcetk` that takes in a `boolean` that indicates whether to use native networking or current `xhr`. Using native stack would be the default and would call into the plugin. If not, the current way of doing things works. I'm thinking `forcetk` would still be the entry point for a hybrid app, as opposed to the plugin (the way it is currently).

wmathurin commented:
Sounds good.

bhariharan commented on libs/forcetk.mobilesdk.js:
```diff
@@ -299,56 +316,61 @@ if (forcetk.Client === undefined) {
-                    if (typeof error == 'function') {
-                        error(jqXHR, textStatus, errorThrown);
+        if (this.useNativeNetworking) {
+            cordova.require("com.salesforce.plugin.network").sendRequest('/services/data', path, callback, error, method, payload, headerParams);
```
@wmathurin This is what I'm proposing. A simple `if-else`. The `else` block is the old implementation with no changes.

bhariharan commented on libs/forcetk.mobilesdk.js:
```diff
@@ -73,13 +73,30 @@ if (forcetk.Client === undefined) {
+     * @param useNativeNetworking True - if native network stack should be used, False - otherwise.
+     * @constructor
+     */
+    forcetk.Client = function(clientId, loginUrl, proxyUrl, authCallback, useNativeNetworking) {
```
Extra optional argument to `useNativeNetworking`. Defaults to `true`. Set to `false` for `xhr`.

wmathurin replied:
Should it default to false instead?
That way, old code that doesn't specify the new parameter would behave like before.

wmathurin replied:
Or if you want default on, then the parameter should be called useXhrNetworking (typically undefined is treated like false).

wmathurin replied:
When you use native networking, are any of the other params of any use: authCallback is useless, what about clientId, loginUrl and proxyUrl?

May be we need two constructors:
```javascript
// Web client
forcetk.Client = function(clientId, loginUrl, proxyUrl, authCallback) {
}
// App client (i.e. in container)
forcetk.appClient = function() {
}
```

bhariharan replied:
They're all pretty useless for the native use case, but I thought I'd keep the API signature the same, for minimal divergence from the original `forcetk`, although I don't know if that's a factor anymore.

bhariharan commented:
Should be good to go now.

bhariharan commented:
@wmathurin Addressed your feedback [here](https://github.com/forcedotcom/SalesforceMobileSDK-Shared/pull/390/commits/d444640d3db37ab680be68ba6b9bfbf9c05091a2).

bhariharan commented:
@wmathurin Good to merge this?

wmathurin commented:
Yep looks good
